### ISSUE TITLE
Dart does not throw error when analyzer fails with "type_argument_not_matching_bounds" error for functions

### ISSUE THREAD

**Issue by iarkh** [dart-lang/sdk#33626]
Dart SDK Version: 2.0.0-dev.64.1
OS: Windows 10 (64 bit)

The following test example tries to create a `Function` object with incorrect type parameter. It produces type_argument_not_matching_bound in analyzer and passes with dart:
``` dart
class A {}
typedef F<X extends A> = X Function();

class B {}
B tryme() {}

main() {
  F<B> f = tryme;
}
```
Sample output is:
```
$> dart --version
Dart VM version: 2.0.0-dev.64.1 (Thu Jun 21 08:47:55 2018 +0200) on "windows_x64"

$> dartanalyzer test.dart
Analyzing test.dart...
  hint - This function declares a return type of 'B', but doesn't end with a return statement at test.dart:5:1 - missing_return
  hint - The value of the local variable 'f' isn't used at test.dart:8:8 - unused_local_variable
2 hints found.

$> dart test.dart

$>
```
I believe dart should throw error here.

Please note that there is are similar issues about incorrect class parameters and mixins: #33585, #33624.

**Comment by anders-sandholm** [comment]
@eernstg

**Comment by eernstg** [comment]
I agree that the VM should emit a compile-time error for that bounds violation; given that #33308 deals with that topic in one place I'll add a comment there that this issue contains additional examples for that topic.

But the analyzer also fails to emit a compile-time error, and it is now the purpose of this issue to get that fixed (relying on #33308 for fixes in the VM).

**Comment by srawlins** [comment]
Analyzer now reports an error here:

> error • 'B' doesn't extend 'A'. • a.dart:8:5 • type_argument_not_matching_bounds
